### PR TITLE
Fixed drag & drop bypassing callbacks

### PR DESCRIPTION
**Pull request by MineWitherMC**
Currently, drag-n-dropping item bypasses callbacks such as `allow_metadata_inventory_put`. See #4321 for more information.
This PR fixes that.

Test code: [here](https://gist.github.com/MineWitherMC/883f70330ebbfb7811c34830104aaeb2)

Before PR:
```
2018-05-25 14:56:12: [Server]: Fired allow_take
2018-05-25 14:56:12: [Server]: Fired take -- these 2 messages are for taking item from Nodemeta, replacing it with players' one
2018-05-25 14:56:14: [Server]: Fired allow_put
2018-05-25 14:56:14: [Server]: Fired put -- these are for opposite
```

After PR:
```
2018-05-25 15:01:12: [Server]: Fired allow_put
2018-05-25 15:01:12: [Server]: Fired allow_take
2018-05-25 15:01:12: [Server]: Fired put
2018-05-25 15:01:12: [Server]: Fired take -- these are for taking item
2018-05-25 15:01:04: [Server]: Fired allow_take
2018-05-25 15:01:04: [Server]: Fired allow_put
2018-05-25 15:01:04: [Server]: Fired take
2018-05-25 15:01:04: [Server]: Fired put -- these are for putting item
```

Todo:

- [ ] Merging stacks (e.g., one inv has max stack size of 10, and has 10 items. Second has max of 5, and has 0 items. Moving items from first one should leave 5 there and put 5 to the second one)
- [ ] Documentation?
- [ ] Rollback support? I didn't get how does it work, will look into it later
- [ ] Rebase, this fork was created like month ago (again, how do I rebase?)

Edit: fixed styling

### CODE REVIEW

**Comment by MineWitherMC** [comment]
Could anyone add [WIP] label here? It is not finished for now (as seen above)

File: src/inventorymanager.cpp [hunk 1]
```diff
@@ -248,6 +248,12 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
 		to_inv.type == InventoryLocation::PLAYER &&
 		from_inv.name == to_inv.name);
 
+	/*
+		Fix Drag & Drop
+	*/
+
+	ItemStack replaced = list_to->getItem(to_i);
+	bool fix_replacements = !replaced.empty();
```
Comment (sfan5) [inline src/inventorymanager.cpp]: bad variable name

File: src/inventorymanager.cpp [hunk 2]
```diff
@@ -272,14 +280,26 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
 	} else {
 		// Destination is detached
 		if (to_inv.type == InventoryLocation::DETACHED) {
-			ItemStack src_item = list_from->getItem(from_i);
+			if (fix_replacements) {
+				int count = PLAYER_TO_SA(player)->detached_inventory_AllowTake(
+						to_inv.name, to_list, to_i, replaced, player);
+				if (replaced.count > count)
+					return;
```
Comment (sfan5) [inline src/inventorymanager.cpp]: what's this for?

File: src/inventorymanager.cpp (hunk 1 above)
Comment (MineWitherMC) [reply src/inventorymanager.cpp]: I know. Would `dst_item` and `dst_item_replaced`(for ItemStack and boolean respectively) be better?

File: src/inventorymanager.cpp (hunk 2 above)
Comment (MineWitherMC) [reply src/inventorymanager.cpp]: This piece checks whether it is possible to take items out of Detached Inventory (e.g, let Detached inventory be put-only, and have item A. Then putting item B on top of A would return A to player without this)

File: src/inventorymanager.cpp (hunk 1 above)
Comment (sfan5) [reply src/inventorymanager.cpp]: I would name it `is_being_replaced`, not too sure about your suggestions but it's probably fine

**Comment by stujones11** [comment]
Good to see someone finally looking into this, it has caused many problems for my armor mod which relies on these inventory callbacks & methods. I will be happy to test this once the conflicts are resolved.

**Comment by paramat** [comment]
@MineWitherMC this seems abandoned? Since it is a bugfix we could mark this as 'adoption needed'.

**Comment by SmallJoker** [comment]
@MineWitherMC Rebased.

**Comment by paramat** [comment]
Seems abandoned. Marked as as 'adoption needed' as it is a bugfix that may be desired and because the author does not know how to rebase.

**Pull request closed without merging**